[Site Isolation] Reduce code duplication when sending IPC to a web page

charliewolfe opened the pull request:
#### 7d27cdd3e980f1d2de99a9525eee9efba8929a07
<pre>
[Site Isolation] Reduce code duplication when sending IPC to a web page
<a href="https://bugs.webkit.org/show_bug.cgi?id=264788">https://bugs.webkit.org/show_bug.cgi?id=264788</a>
<a href="https://rdar.apple.com/118368823">rdar://118368823</a>

Reviewed by Alex Christensen.

With site isolation we often need to check if a RemotePageProxy exists for a given frame before sending
IPC to the process hosting the frame. We should have a helper function instead of duplicating this logic
throughout WebPageProxy.

* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::sendMouseEvent): Deleted.
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::performDragControllerAction):
(WebKit::WebPageProxy::dragEnded):
(WebKit::WebPageProxy::handleMouseEventReply):
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::sendKeyEvent):
(WebKit::WebPageProxy::runJavaScriptInFrameInScriptWorld):
(WebKit::WebPageProxy::dispatchLoadEventToFrameOwnerElement):
(WebKit::WebPageProxy::sendToWebPage):
(WebKit::WebPageProxy::sendToProcessContainingFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: <a href="https://commits.webkit.org/270723@main">https://commits.webkit.org/270723@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8066ba204b108237a00395bc05faa9acb6c0d273

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26240 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4850 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/27508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28336 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/24023 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/6635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/2271 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26496 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/6635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/27508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28914 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/6635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/27508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29592 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/6635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/27508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27480 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/3338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/2271 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/4752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/27508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/3809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3376 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/3663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -2849,6 +2849,10 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 
     void dispatchLoadEventToFrameOwnerElement(WebCore::FrameIdentifier);
 
+    template<typename F> void sendToWebPage(const std::optional<WebCore::FrameIdentifier>&, F&&);
```
This can be private, right?

charliewolfe replied:
The function is private? Or do you mean these should be in their own class?

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13276,6 +13225,43 @@ Ref<WebBackForwardList> WebPageProxy::protectedBackForwardList() const
+            }
+        } else {
+            ASSERT_NOT_REACHED();
+            return;
```
I don't think we would want to return in this case.

charliewolfe replied:
Returning here matches existing behavior, but I agree maybe we shouldn't. I'll remove it.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13276,6 +13225,43 @@ Ref<WebBackForwardList> WebPageProxy::protectedBackForwardList() const
+}
+
+template<typename M, typename C>
+void WebPageProxy::sendAsyncMessageToWebPage(const std::optional<FrameIdentifier>& frameID, M&& message, C&& completionHandler)
```
I think all the callers of this have a non-optional FrameIdentifier.  Is that right?  Otherwise, this should probably not be a const reference because it can just be in registers so there's no reason to pass only a pointer.

charliewolfe replied:
The calls in `dragEnded()` and `performDragControllerAction()` have an optional FrameIdentifier.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -13276,6 +13225,41 @@ Ref<WebBackForwardList> WebPageProxy::protectedBackForwardList() const
+    if (frameID) {
+        if (RefPtr frame = WebFrameProxy::webFrame(*frameID)) {
+            if (RefPtr remotePageProxy = frame->remotePageProxy()) {
+                sendFunction(remotePageProxy.get());
```
Would it work to call sendFunction(*remotePageProxy) and sendFunction(*this) and have the lambda parameter be auto& instead of auto*?

charliewolfe committed:
[Site Isolation] Reduce code duplication when sending IPC to a web page
https://bugs.webkit.org/show_bug.cgi?id=264788
rdar://118368823

Reviewed by Alex Christensen.

With site isolation we often need to check if a RemotePageProxy exists for a given frame before sending
IPC to the process hosting the frame. We should have a helper function instead of duplicating this logic
throughout WebPageProxy.

* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::sendMouseEvent): Deleted.
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::performDragControllerAction):
(WebKit::WebPageProxy::dragEnded):
(WebKit::WebPageProxy::handleMouseEventReply):
(WebKit::WebPageProxy::sendMouseEvent):
(WebKit::WebPageProxy::sendKeyEvent):
(WebKit::WebPageProxy::runJavaScriptInFrameInScriptWorld):
(WebKit::WebPageProxy::dispatchLoadEventToFrameOwnerElement):
(WebKit::WebPageProxy::sendToWebPage):
(WebKit::WebPageProxy::sendToProcessContainingFrame):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: https://commits.webkit.org/270723@main
